File: Application/src/main/java/com/shareyourproxy/api/rx/command/SaveGroupChannelsCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;
import android.os.Parcelable;

import com.shareyourproxy.api.domain.model.Group;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;
import com.shareyourproxy.app.EditGroupChannelsActivity.GroupEditType;

import java.util.HashSet;
import java.util.List;

import static com.shareyourproxy.api.rx.RxGroupChannelSync.updateGroupChannels;

/**
 * Save channels associated with a group.
 */
public class SaveGroupChannelsCommand extends BaseCommand {
    public static final Parcelable.Creator<SaveGroupChannelsCommand> CREATOR =
        new Parcelable.Creator<SaveGroupChannelsCommand>() {
            @Override
            public SaveGroupChannelsCommand createFromParcel(Parcel in) {
                return new SaveGroupChannelsCommand(in);
            }

            @Override
            public SaveGroupChannelsCommand[] newArray(int size) {
                return new SaveGroupChannelsCommand[size];
            }
        };
    private final static java.lang.ClassLoader CL =
        SaveGroupChannelsCommand.class.getClassLoader();

    private final User user;
    private final String newTitle;
    private final Group group;
    private final HashSet<String> channels;
    private final GroupEditType groupEditType;


    public SaveGroupChannelsCommand(User user, String newTitle, Group group,
        HashSet<String> channels, GroupEditType groupEditType) {
        this.user = user;
        this.newTitle = newTitle;
        this.group = group;
        this.channels = channels;
        this.groupEditType = groupEditType;
    }

    private SaveGroupChannelsCommand(Parcel in) {
        this((User) in.readValue(CL), (String) in.readValue(CL),
            (Group) in.readValue(CL), (HashSet<String>) in.readValue(CL),
            (GroupEditType) in.readValue(CL));
    }

    @Override
    public EventCallback execute(Service service) {
        return updateGroupChannels(service, user, newTitle, group, channels, groupEditType);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(user);
        dest.writeValue(newTitle);
        dest.writeValue(group);
        dest.writeValue(channels);
        dest.writeValue(groupEditType);
    }

}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/SaveGroupChannelsCommand.kt
package com.shareyourproxy.api.rx.command

import android.content.Context
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.model.Group
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxGroupChannelSync.updateGroupChannels
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback
import com.shareyourproxy.util.Enumerations
import java.util.*

/**
 * Save channels associated with a group.
 */
internal class SaveGroupChannelsCommand(private val user: User, private val newTitle: String, private val group: Group, private val channels: HashSet<String>, private val groupEditType: Enumerations.GroupEditType) : BaseCommand() {
    @Suppress("UNCHECKED_CAST")
    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as String, parcel.readValue(CL) as Group, parcel.readValue(CL) as HashSet<String>, parcel.readValue(CL) as Enumerations.GroupEditType)

    override fun execute(context: Context): EventCallback {
        return updateGroupChannels(context, user, newTitle, group, channels, groupEditType)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(newTitle)
        dest.writeValue(group)
        dest.writeValue(channels)
        dest.writeValue(groupEditType)
    }

    companion object {
        private val CL = SaveGroupChannelsCommand::class.java.classLoader
        @JvmField val CREATOR: Parcelable.Creator<SaveGroupChannelsCommand> = object : Parcelable.Creator<SaveGroupChannelsCommand> {
            override fun createFromParcel(parcel: Parcel): SaveGroupChannelsCommand {
                return SaveGroupChannelsCommand(parcel)
            }

            override fun newArray(size: Int): Array<SaveGroupChannelsCommand?> {
                return arrayOfNulls(size)
            }
        }
    }
}
